logging/include/cocaine/logging/metafilter.hpp

```diff
@@ -0,0 +1,83 @@
+/*
+    Copyright (c) 2011-2015 Anton Matveenko <antmat@yandex-team.ru>
+    Copyright (c) 2011-2015 Other contributors as noted in the AUTHORS file.
+
+    This file is part of Cocaine.
+
+    Cocaine is free software; you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    Cocaine is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include "cocaine/filter.hpp"
+#include <cocaine/repository.hpp>
+
+namespace cocaine { namespace logging {
+
+class metafilter_t {
+public:
+    metafilter_t(std::string name_) :
+        name(std::move(name_))
+    {}
+
+    bool apply(const blackhole::attribute::combined_view_t&) {
+        for(const auto& filter: filters) {
+            if(!filter.apply()) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    void add_filter(filter_info_t filter) {
+        filters.push_back(std::move(filter));
+    }
+private:
+    std::vector<filter_info_t> filters;
+};
+
+template<>
+struct category_traits<metafilter_t> {
+    typedef std::shared_ptr<metafilter_t> ptr_type;
+
+    struct factory_type: public api::basic_factory<metafilter_t> {
+        virtual
+        ptr_type
+        get(const std::string& name) = 0;
+    };
+
+    template<class T>
+    struct default_factory: public factory_type {
+        virtual
+        ptr_type
+        get(const std::string& name) {
+            return instances.apply([&](std::map<std::string, std::weak_ptr<storage_t>>& instances) {
```

Comment:
Yeah, I've been thinking about it for a long time. We essentially have two types of factories now – the one that creates a new instance every call and returns `std::unique_ptr<T>` and the one that keeps a map of `std::weak_ptr<T>` and returns a cached instance if it's still in use somewhere.